src/Common/ZooKeeper/ZooKeeperLoadBalancer.cpp

```diff
@@ -0,0 +1,159 @@
+    std::lock_guard lock{mutex};
+    args = args_;
+    zk_log = std::move(zk_log_);
+    log = &Poco::Logger::get("ZooKeeperLoadBalancer");
+    for (size_t i = 0; i < args_.hosts.size(); ++i)
+    {
+        HostInfo host_info;
+        auto [address, secure] =  parseForSocketAddress(args_.hosts[i]);
+        host_info.address = address;
+        host_info.original_index = i;
+        host_info.secure = secure;
+        host_info_list.push_back(host_info);
+    }
+}
+
+void ZooKeeperLoadBalancer::setBetterKeeperHostUpdater(BetterKeeperHostUpdater handler)
+{
+    std::lock_guard lock{mutex};
+    better_keeper_host_updater = handler;
+}
+
+
+[[noreturn]] void throwWhenNoHostAvailable(bool dns_error_occurred)
+{
+    if (dns_error_occurred)
+        throw zkutil::KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot resolve any of provided ZooKeeper hosts due to DNS error");
+    else
+        throw zkutil::KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot use any of provided ZooKeeper nodes");
+}
+
+
+void ZooKeeperLoadBalancer::shuffleHosts()
+{
+    std::function<Priority(size_t index)> get_priority = args.get_priority_load_balancing.getPriorityFunc(args.get_priority_load_balancing.load_balancing, 0, args.hosts.size());
+    for (size_t i = 0; i < host_info_list.size(); ++i)
+    {
+        if (get_priority)
+            host_info_list[i].priority = get_priority(i);
+        host_info_list[i].randomize();
+    }
+
+    ::sort(host_info_list.begin(), host_info_list.end(), HostInfo::compare);
+}
+
+void ZooKeeperLoadBalancer::recordKeeperHostError(UInt8 original_index)
+{
+    std::lock_guard lock{mutex};
+    for (auto & host_info : host_info_list)
+    {
+        if (host_info.original_index == original_index)
+        {
+            LOG_DEBUG(log, "Load Balancer records a failure on host {}, original index {}", host_info.address, static_cast<UInt32>(original_index));
+            // TODO: add the actual implementation to lower the priority/scoring of the host.
+            break;
+        }
+    }
+    LOG_ERROR(log, "Does not find host to record the failure with original index {}", static_cast<UInt32>(original_index));
+}
+
+std::unique_ptr<Coordination::ZooKeeper> ZooKeeperLoadBalancer::createClient()
+{
+    std::lock_guard lock{mutex};
+    shuffleHosts();
+    bool dns_error_occurred = false;
+    for (size_t i = 0; i < host_info_list.size(); ++i)
+    {
+        if (!isKeeperHostDNSAvailable(log, args.hosts[i], dns_error_occurred))
+            continue;
+
+        LOG_INFO(log, "Connecting to ZooKeeper host {}, number of attempted hosts {}", host_info_list[i].address, i+1);
+        try
+        {
+            auto client  = std::make_unique<Coordination::ZooKeeper>(host_info_list[i].toZooKeeperNode(), args, zk_log);
+            // Non optimal case: we connected to a keeper host in different availability zone, so set a session deadline.
```

Comment:
Do you introduce the local az policy in that PR?

Reply:
Updated the comment.